### PR TITLE
Refactor history events replication apply logic

### PR DESCRIPTION
**Pull request by xwduan**
## What changed?
Refactor history events replication apply logic

## Why?

1. To support events import
2. As a building block to handle migration back and forth scenario

## How did you test it?

unit test
## Potential risks

no risk, not wired into current logic
## Is hotfix candidate?

no

### CODE REVIEW

File: common/persistence/versionhistory/version_history.go [hunk 1]
```diff
@@ -176,6 +176,19 @@ func IsVersionHistoryItemsInSameBranch(versionHistoryItemsA []*historyspb.Versio
 	return lowestCommonAncestor.Equal(aLastItem) || lowestCommonAncestor.Equal(bLastItem)
 }
 
+func ParseVersionHistoryToPastAndFuture(
```
Comment (yux0) [inline common/persistence/versionhistory/version_history.go:179]: SplitVersionHistoryByLastItem?

File: service/history/replication/eventhandler/history_events_handler.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	}
+	lastPastBatchIndex := -1
+	for index, batch := range eventsBatches {
+		if batch[len(batch)-1].EventId == lastPastEventId {
+			lastPastBatchIndex = index
+			break
+		}
+	}
+	if lastPastBatchIndex == -1 {
+		return nil, nil, serviceerror.NewInvalidArgument("No boundary events found") // if this happens, means the events are not consecutive and we have bug somewhere
```
Comment (yux0) [inline service/history/replication/eventhandler/history_events_handler.go:153]: InternalServiceError?

File: service/history/replication/eventhandler/history_events_handler.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+			baseExecutionInfo *workflowpb.BaseExecutionInfo,
+			versionHistoryItems []*historyspb.VersionHistoryItem,
+			historyEvents [][]*historypb.HistoryEvent,
+			newEvents []*historypb.HistoryEvent,
+		) error
+	}
+
+	historyEventsHandlerImpl struct {
+		replication.ProcessToolBox
+		PastEventsHandler
```
Comment (yux0) [inline service/history/replication/eventhandler/history_events_handler.go]: I feel the past/future is not good wording here. It is hard to understand. could it just be import/replicate? Or LocalEventHandler and RemoteEventHandler?

Comment (xwduan) [reply service/history/replication/eventhandler/history_events_handler.go]: Yeah, I like "LocalEventHandler and RemoteEventHandler". Will change it in next revision.

Comment (xwduan) [reply service/history/replication/eventhandler/history_events_handler.go]: I actually asked GPT to come up with a name, Local and Remote is one of them!

File: common/persistence/versionhistory/version_history.go [hunk 4]
```diff
@@ -176,6 +176,19 @@ func IsVersionHistoryItemsInSameBranch(versionHistoryItemsA []*historyspb.Versio
 	return lowestCommonAncestor.Equal(aLastItem) || lowestCommonAncestor.Equal(bLastItem)
 }
 
+func SplitVersionHistoryByLastLocalGeneratedItem(
+	versionHistoryItems []*historyspb.VersionHistoryItem,
+	initialFailoverVersion int64,
+	failoverVersionIncrement int64,
+) ([]*historyspb.VersionHistoryItem, []*historyspb.VersionHistoryItem) {
```
Comment (yux0) [inline common/persistence/versionhistory/version_history.go:183]: nit: add the para name to better understand what are the return items?

File: common/persistence/versionhistory/version_history.go [hunk 5]
```diff
@@ -176,6 +176,19 @@ func IsVersionHistoryItemsInSameBranch(versionHistoryItemsA []*historyspb.Versio
...
+	for i := len(versionHistoryItems) - 1; i >= 0; i-- {
+		if versionHistoryItems[i].Version%failoverVersionIncrement == initialFailoverVersion {
+			return versionHistoryItems[:i+1], versionHistoryItems[i+1:]
```
Comment (yux0) [inline common/persistence/versionhistory/version_history.go:186]: just curious why +1, it seems like it doesn't match the function name?

**Review by yux0 — APPROVED** [review]

File: common/persistence/versionhistory/version_history.go (hunk 4 above)
Comment (xwduan) [reply common/persistence/versionhistory/version_history.go:183]: Sure. Will do.

File: common/persistence/versionhistory/version_history.go (hunk 5 above)
Comment (xwduan) [reply common/persistence/versionhistory/version_history.go:186]: It iterate backwards to find the last local generated item. The `versionHistoryItems[:i+1]` is `[inclusive,exclusive]` while I need to include the last local generated item.

Comment (yux0) [reply common/persistence/versionhistory/version_history.go:186]: Thank you. I understand it now

**Pull request merged**